server/src/main/java/org/elasticsearch/indices/recovery/PeerRecoveryTargetService.java

```diff
@@ -366,13 +367,11 @@ private void performTranslogOps(final RecoveryTranslogOperationsRequest request,
                 observer.waitForNextChange(new ClusterStateObserver.Listener() {
                     @Override
                     public void onNewClusterState(ClusterState state) {
-                        try {
+                        threadPool.generic().execute(ActionRunnable.wrap(listener, l -> {
```

Comment:
Could also go for the write pool here (as suggested in https://github.com/elastic/elasticsearch/pull/58018) but I figured I wouldn't change that in this PR.

Reply:
+1